File: DSA/src/Questions/Matrix/DiagonalTraversal.java
package Questions.Matrix;

/*
https://leetcode.com/problems/diagonal-traverse/description/
 Given an m x n matrix mat, return an array of all the elements of the array in a diagonal order.

 

Example 1:


Input: mat = [[1,2,3],[4,5,6],[7,8,9]]
Output: [1,2,4,7,5,3,6,8,9]
Example 2:

Input: mat = [[1,2],[3,4]]
Output: [1,2,3,4]
 

Constraints:

m == mat.length
n == mat[i].length
1 <= m, n <= 104
1 <= m * n <= 104
-105 <= mat[i][j] <= 105
Time Complexity: O(m * n)
Space Complexity: O(1)
 */
public class DiagonalTraversal {
    public int[] findDiagonalOrder(int[][] mat) {
        if (mat == null || mat.length == 0)
            return new int[0];

        int m = mat.length;
        int n = mat[0].length;
        int[] result = new int[m * n];
        int r = 0, c = 0, d = 1;

        for (int i = 0; i < m * n; i++) {
            result[i] = mat[r][c];
            // Move up-right
            if (d == 1) {
                if (c == n - 1) {
                    r++;
                    d = -1;
                } else if (r == 0) {
                    c++;
                    d = -1;
                } else {
                    r--;
                    c++;
                }
            }
            // Move down-left
            else {
                if (r == m - 1) {
                    c++;
                    d = 1;
                } else if (c == 0) {
                    r++;
                    d = 1;
                } else {
                    r++;
                    c--;
                }
            }
        }

        return result;
    }
    public static void main(String[] args) {
        DiagonalTraversal diagonalTraversal = new DiagonalTraversal();
        int[][] mat = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        int[] result = diagonalTraversal.findDiagonalOrder(mat);
        for (int i : result) {
            System.out.print(i + " ");
        }
    }
}

File: DSA/Python/Questions/Matrix/DiagonalTraversal.py
"""
# Question: Diagonal Traverse
# Link: https://leetcode.com/problems/diagonal-traverse/

# Time Complexity: O(m*n)
# Space Complexity: O(m*n)

# Algorithm:
# 1. Track direction (up/down)
# 2. Process each diagonal
# 3. Handle boundary conditions
# 4. Return traversal result
"""


class DiagonalTraversal:
    def find_diagonal_order(self, matrix: list[list[int]]) -> list[int]:
        if not matrix or not matrix[0]:
            return []

        m, n = len(matrix), len(matrix[0])
        result = []
        row = col = 0
        direction = 1  # 1 for up, -1 for down

        while len(result) < m * n:
            result.append(matrix[row][col])

            if direction == 1:
                if col == n - 1:
                    row += 1
                    direction = -1
                elif row == 0:
                    col += 1
                    direction = -1
                else:
                    row -= 1
                    col += 1
            else:
                if row == m - 1:
                    col += 1
                    direction = 1
                elif col == 0:
                    row += 1
                    direction = 1
                else:
                    row += 1
                    col -= 1

        return result


def main():
    solution = DiagonalTraversal()
    matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    print(solution.find_diagonal_order(matrix))


if __name__ == "__main__":
    main()
